lua-mode.el

```diff
@@ -1814,6 +1817,105 @@ If `lua-process' is nil or dead, start a new process first."
+               "        table.insert(matches, table.concat(parts, '.'))"
+               "      end"
+               "    end"
+               "    return matches"
+               "  else" ; more segments of input remain; descend into ctx table
+               "    local token1 = table.remove(input_parts, 1)"
+               "    table.insert(prefixes, first_part)"
+               "    return cpl_for(input_parts, ctx[token1], prefixes)"
+               "  end"
+               "end"
+
+               ;; this is a table of the segments of the input
+               "local input = {"
+               ,@(mapcar (apply-partially 'format "'%s',")
+                         (split-string expr "\\.")) "}"
+               ;; spit it out to a file; lua-mode can't send data back to emacs
+               ,(format "local f = io.open('%s', 'w')" out-file)
+               "for _,l in ipairs(cpl_for(input, top_ctx, {})) do"
+               "  f:write(l .. string.char(10))"
+               "end"
+               "f:close()"
+               "end") "\n"))
+
+(defvar lua-local-require-completions nil
+  "During completion, scan file for local require calls for context.
+
+Defaults to nil because this will cause code to be loaded during completion.
+Loading arbitrary code can have unexpected side-effects, so use with caution.")
+
+(defvar lua-local-require-regexp
+  "^local\\s-+\\(\\w+\\)\\s-*=\\s-*require[( ]+\\([^ )\n]+\\)"
+  "A regexp to match lines where a library is required and put in a local.")
+
+(defun lua-local-libs ()
+  "Find all modules loaded with require which are stored in locals.
+
+Returns a list of lists where the car is the local name and the cadr
+is the string that is passed to require."
+  (save-excursion
+    (when lua-local-require-completions
+      (let ((libs nil))
+        (goto-char (point-min))
+        ;; find each match of "local x = require y" and save for later
+        (while (search-forward-regexp lua-local-require-regexp nil t)
+          (add-to-list 'libs (list (match-string-no-properties 1)
+                                   (match-string-no-properties 2))))
+        libs))))
+
+(defun lua-complete-function ()
+  "Completion function for `completion-at-point-functions'.
+
+Queries current lua subprocess for possible completions."
+  (let* ((start-of-expr (save-excursion
+                          (search-backward-regexp "[^\.a-zA-Z0-9_]") (point)))
+         (start-of-token (save-excursion (when (symbol-at-point)
+                                           (backward-word)) (point)))
+         (expr (buffer-substring-no-properties (1+ start-of-expr) (point)))
+         (libs (lua-local-libs))
+         (file (make-temp-file "lua-completions-")))
+    (lua-send-string (lua-completion-string-for expr libs file))
+    (sit-for 0.1)
+    (list start-of-token (point)
+          (when (file-exists-p file)
+            (with-temp-buffer
+              (insert-file-contents file)
+              (delete-file file)
```

Comment:
Is it intentional that we don't delete the file if something goes south? Should we wrap it into `unwind-protect`?